internal/therealproxy/client.go

```diff
@@ -16,20 +21,110 @@ var Log = logging.MustGetLogger("therealproxy")
 type Client struct {
 	session  *yamux.Session
 	listener net.Listener
+	app      *app.App
+	addr     routing.Addr
 }
 
 // NewClient constructs a new Client.
-func NewClient(conn net.Conn) (*Client, error) {
+func NewClient(lis net.Listener, app *app.App, addr routing.Addr) (*Client, error) {
+	c := &Client{
+		listener: lis,
+		app:      app,
+		addr:     addr,
+	}
+	if err := c.connect(); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *Client) connect() error {
+	r := netutil.NewRetrier(time.Second, 0, 1)
+
+	var conn net.Conn
+	err := r.Do(func() error {
+		var err error
+		conn, err = c.app.Dial(c.addr)
+		return err
+	})
+	if err != nil {
+		return fmt.Errorf("failed to dial to a server: %v", err)
+	}
+
 	session, err := yamux.Client(conn, nil)
 	if err != nil {
-		return nil, fmt.Errorf("yamux: %s", err)
+		return fmt.Errorf("failed to create client: %s", err)
 	}
 
-	return &Client{session: session}, nil
+	c.session = session
+
+	return nil
+}
+
+// Serve proxies incoming connection to a remote proxy server.
+func (c *Client) Serve() error {
+	for {
+		conn, err := c.listener.Accept()
+		if err != nil {
+			return fmt.Errorf("accept: %s", err)
+		}
+
+		stream := c.createStream()
+		c.handleStream(conn, stream)
+	}
+}
+
+func (c *Client) createStream() net.Conn {
+	for {
+		stream, err := c.session.Open()
+		if err == nil {
+			return stream
+		}
+
+		Log.Warnf("Failed to open yamux session: %v", err)
+
+		delay := 1 * time.Second
+		Log.Warnf("Restarting in %v", delay)
+		time.Sleep(delay)
+
+		if err := c.connect(); err != nil {
+			Log.Warnf("Failed to reconnect, trying again")
+		}
+	}
+}
+
+func (c *Client) handleStream(in, out net.Conn) {
+	go func() {
+		errCh := make(chan error, 2)
+		go func() {
+			_, err := io.Copy(out, in)
+			errCh <- err
+		}()
+
+		go func() {
+			_, err := io.Copy(in, out)
+			errCh <- err
+		}()
+
+		for err := range errCh {
+			if err := in.Close(); err != nil {
+				Log.WithError(err).Warn("Failed to close connection")
+			}
+			if err := out.Close(); err != nil {
+				Log.WithError(err).Warn("Failed to close stream")
+			}
+
+			if err != nil {
+				Log.Error("Copy error:", err)
+			}
+		}
+	}()
 }
 
-// ListenAndServe start tcp listener on addr and proxies incoming
+// ListenAndServe starts tcp listener on addr and proxies incoming
 // connection to a remote proxy server.
+// TODO: get rid of it
 func (c *Client) ListenAndServe(addr string) error {
```

Comment:
I think we should get rid of this. Since we're passing listener to the client's contructor and have a separate `Serve` func, this `ListenAndServe` makes things confusing

Reply:
@Darkren Good catch. But not sure if it's worth for milestone 1 because the new app is going to be integrated into milestone 2, so the proxy client will require modifications anyway.